client/src/main/java/io/pravega/client/stream/impl/StreamSegments.java

```diff
@@ -63,64 +74,110 @@ public Segment getSegmentForKey(String key) {
+        for (SegmentWithRange seg : segments.values()) {
+            result.add(seg.getSegment());
+        }
+        return result;
     }
     
-    public StreamSegments withReplacementRange(Segment replacedSegment, StreamSegmentsWithPredecessors replacementRanges) {
-        Preconditions.checkState(segments.containsValue(replacedSegment), "Segment to be replaced should be present in the segment list");
-        verifyReplacementRange(replacementRanges);
-        NavigableMap<Double, Segment> result = new TreeMap<>();
+    public StreamSegments withReplacementRange(Segment segment, StreamSegmentsWithPredecessors replacementRanges) {
+        SegmentWithRange replacedSegment = findReplacedSegment(segment);
+        verifyReplacementRange(replacedSegment, replacementRanges);
+        NavigableMap<Double, SegmentWithRange> result = new TreeMap<>();
         Map<Long, List<SegmentWithRange>> replacedRanges = replacementRanges.getReplacementRanges();
-        List<SegmentWithRange> replacements = replacedRanges.get(replacedSegment.getSegmentId());
-        Preconditions.checkNotNull(replacements, "Empty set of replacements for: {}", replacedSegment.getSegmentId());
+        List<SegmentWithRange> replacements = replacedRanges.get(segment.getSegmentId());
+        Preconditions.checkNotNull(replacements, "Empty set of replacements for: {}", segment.getSegmentId());
         replacements.sort(Comparator.comparingDouble(SegmentWithRange::getHigh).reversed());
-        Segment lastSegmentValue = null;
-        for (Entry<Double, Segment> existingEntry : segments.descendingMap().entrySet()) { // iterate from the highest key.
-            final Segment existingSegment = existingEntry.getValue();
-            if (existingSegment.equals(lastSegmentValue)) {
-                // last value was the same segment, it can be consolidated.
-                continue;
-            }
+        for (Entry<Double, SegmentWithRange> existingEntry : segments.descendingMap().entrySet()) { // iterate from the highest key.
+            final SegmentWithRange existingSegment = existingEntry.getValue();
             if (existingSegment.equals(replacedSegment)) { // Segment needs to be replaced.
                 // Invariant: The successor segment(s)'s range should be limited to the replaced segment's range, thereby
                 // ensuring that newer writes to the successor(s) happen only for the replaced segment's range.
-                for (SegmentWithRange segmentWithRange : replacements) {                        
-                    result.put(Math.min(segmentWithRange.getHigh(), existingEntry.getKey()), segmentWithRange.getSegment());
+                for (SegmentWithRange segmentWithRange : replacements) {
+                    Double lowerBound = segments.lowerKey(existingEntry.getKey()); // Used to skip over items not in the clients view yet.
+                    if (lowerBound == null || segmentWithRange.getHigh() >= lowerBound) { 
```

Comment:
Did not understand this change from code inspection. 
lowerBound always points to the next highest key after existingSegment. Isn't `segmentWithRange.getHigh()` always greater than `lowerBound` ?

Reply:
It is not for example:
If segments 0 and 1 merge into 2. 
Later 2 splits into 3,4,5.
The client encounters EOS on segment 1.
Then it's map contain 0 and 2 where 2 is only occupying the range that 1 used to rather than its full span. 
If it then encounters an EOS on 2 before 1 the segment 3 which occupies the lower portion of keyspace shouldn't be added to the map. This if condition excludes it. 
It will eventually be added after EOS is encountered on 0 and then 2 for the second time.

Please suggest phrasing that would make this clear.